Add basic implementation of batch read records mode

igkon opened the pull request:
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

Add submitBatchRecordRead method that polls records according to MAX_POLL_RECORDS_CONFIG parameter. The parameter is 'to' minus 'from'.
Also add corresponding batch methods such as completeRecordReadOperations, submitEach and change submit(List<O> ops...) functions.

Commit:
Add basic implementation of batch read records mode

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

Add submitBatchRecordRead method that polls records according to MAX_POLL_RECORDS_CONFIG parameter. The parameter is 'to' minus 'from'.
Also add corresponding batch methods such as completeRecordReadOperations, submitEach and change submit(List<O> ops...) functions.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
+        consumerConfig.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, String.valueOf(amountOfRecords));
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+        final KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(consumerConfig);
+        consumer.subscribe(Arrays.asList(anyOp.dstPath())); // Could ops have different dstPaths?
```
yes

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
+          O recordOp = ops.get(i);
+          recordOp.startRequest();
+        }
+        val readRecords = consumer.poll(Duration.ZERO);
```
invoke the startRequest and finishRequest before poll for all operations in the range to - from

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
+          recordOp.startRequest();
+        }
+        val readRecords = consumer.poll(Duration.ZERO);
+        for (ConsumerRecord<String, byte[]> record : readRecords) {
```
1. This will create the new iterator instance, please iterate using an index
2. submitCount should be determined here as the actual count of the returned records
3. for each corresponding operation invoke startResponse and finishResponse here

igkon replied:
1. It's impossible because we can't use ConsumerRecords<String, byte[]> as List. We can get list from ConsumerRecords by TopicPartition only.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
+        if (sizesOfReadRecords.size() >= amountOfRecords) {
+          completeRecordReadOperations(ops, from, to, SUCC, sizesOfReadRecords);
+          submitCount = amountOfRecords;
+          consumer.commitSync();
```
why do we need this?

igkon replied:
I think that if we failed to read at least one operation from the list we need to roll back the offset. Am I wrong?

akurilov replied:
but then we could read some records twice, right?

igkon replied:
Hm. Ok then we can commitSync after each correctly read record

igkon replied:
Nope. We can't use it like this. Because we need to commitSync after poll method only. I'll remove manual offset control for now.

akurilov commented on src/main/resources/config-schema-storage-driver-kafka.yml:
```diff
@@ -4,6 +4,8 @@ load:
     timeoutMillis: int
 storage:
   driver:
+    record:
```
I don't think we need the configuration parameter. Just read in the batch mode always

igkon replied:
Ok. Removed

Commit:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-kafka into issue-42-batchReadRecords

Commit:
Fix after review

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

Remove manual offset control for now because we can't commit offset after each record in batch mode. Remove configParameter for batch read mode because it will be set as default. Add method to get all topic names from dstPaths of operations. Also change places of invocations of methods handling the operations.

Commit:
Work with one topic

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

Now all ops from list have one dstPath(). It means that batchRead should work with only one topic for now.

Commit:
fix

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

Commit:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-kafka into issue-42-batchReadRecords
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

AnastasiaShemyakinskaya committed:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-kafka into issue-42-batchReadRecords

AnastasiaShemyakinskaya committed:
Fixed code
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Removed unnecessary lines
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Remove unnecessary lines
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -373,12 +369,15 @@ void consumeRecords(final List<O> recOps) {
         try {
           concurrencyThrottle.acquire();
           var recOp = (DataOperation) null;
+          Set<String> topics = new HashSet<>();
```
please consider using the `val` or `final` here

Commit:
Change Set<String> to val
